### DISCUSSION TITLE
How to save workflow state after each activity execution when workflow created in programtic way.

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#4966] [Q&A]
How to save workflow state after each activity execution when workflow created in programtic way. As of now I am saving workflow state in mongo db. how to retrieve latest state resuming a workflow from a suspended state?
Consider workflow and activities are like:

 public class CustomWorkFlow : IWorkflow
 {
     public void Build(IWorkflowBuilder builder)
     {
         builder
         .StartWith<FirstActivity>()
         .Then<SecondActivity>()
         .Then<ThirdActivity>();
     }
 }

 public class FirstActivity() : Elsa.Services.Activity

 {
     public override async ValueTask<IActivityExecutionResult> ExecuteAsync(Elsa.Services.Models.ActivityExecutionContext context)
     {
         return await Task.FromResult(Done()); //this may vary
     }
 }

 public class SecondActivity() : Elsa.Services.Activity

 {
     public override async ValueTask<IActivityExecutionResult> ExecuteAsync(Elsa.Services.Models.ActivityExecutionContext context)
     {
         return await Task.FromResult(Done());//this may vary
     }

     public override async ValueTask<IActivityExecutionResult> ResumeAsync(Elsa.Services.Models.ActivityExecutionContext context)
     {
         return await Task.FromResult(Done()); //this may vary
     }
 }

 public class ThirdActivity() : Elsa.Services.Activity

 {
     public override async ValueTask<IActivityExecutionResult> ExecuteAsync(Elsa.Services.Models.ActivityExecutionContext context)
     {
         return await Task.FromResult(Suspend()); //this may vary
     }

     public override async ValueTask<IActivityExecutionResult> ResumeAsync(Elsa.Services.Models.ActivityExecutionContext context)
     {
         return await Task.FromResult(Done()); //this may vary
     }

 }

**Comment by glime-ai** [comment]
To save the workflow state after each activity execution when the workflow is created programmatically, you can leverage the persistence features provided by Elsa. Since you are already using MongoDB to save the workflow state, you should ensure that the workflow execution context is being persisted after each activity execution.

In Elsa, the workflow state is automatically persisted when using the `WorkflowRunner` or `IWorkflowInvoker` services to execute workflows. If you need to manually control the persistence of the workflow state, you can use the `IWorkflowInstanceStore` to save the state after each activity execution.

To retrieve the latest state and resume a workflow from a suspended state, you can use the `IWorkflowRunner` or `IWorkflowInvoker` services to resume the workflow by providing the workflow instance ID and the bookmark that indicates where the workflow was suspended.

Here's a high-level example of how you might persist and resume a workflow:

```csharp
// Inject these services where you need to manage workflow execution and persistence.
private readonly IWorkflowRunner _workflowRunner;
private readonly IWorkflowInstanceStore _workflowInstanceStore;

// To execute and persist the workflow state after each activity.
var workflowInstance = await _workflowRunner.RunAsync(customWorkflow);

// To resume a workflow from a suspended state, you need the workflow instance ID and the bookmark.
var workflowInstanceId = workflowInstance.Id; // Get the workflow instance ID.
var bookmark = workflowInstance.Bookmarks.FirstOrDefault(); // Get the bookmark if available.

// Resume the workflow using the instance ID and bookmark.
var result = await _workflowRunner.RunAsync(
    workflowInstance.DefinitionId,
    workflowInstanceId,
    bookmark?.Name,
    bookmark?.Hash
);
```

Please note that the above code is a simplified example. The actual implementation may vary depending on the version of Elsa you are using and the specific requirements of your application.

For more detailed information and examples, you should refer to the official Elsa documentation and the source code examples. Since you are using MongoDB, make sure that the MongoDB provider is correctly configured to handle the persistence of workflow instances.

If you encounter any issues or need further assistance, consider reaching out to the Elsa community on GitHub or Stack Overflow for support.
